### PR TITLE
Gather for CUDA support

### PR DESCRIPTION
**Pull request by yuehhua**

### CODE REVIEW

File: src/gather.jl [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    @inbounds if index <= max_idx
+        j, k = divrem(index-1, max_dims_idx)
+        dims_i = CartesianIndices(dims_size)[k+1]
+        dst[index] = src[Tuple(dims_i)..., idx[j+1]...]
+    end
+    return nothing
+end
+
+function NNlib.gather!(dst::AnyCuArray, src::AnyCuArray, idx::AnyCuArray)
+    dims = NNlib._check_dims(src, dst, idx)
```
Comment (CarloLucibello) [inline src/gather.jl:13]: can you copy the definition over here `_check_dims`? Better not rely on internals as this is a separate library (also the name `_check_dims` is too generic, we should change it to `gather_check_dims`

File: src/gather.jl [hunk 2]
```diff
@@ -0,0 +1,25 @@
+function gather_kernel!(dst, src, idx, max_idx, max_dims_idx, dims_size)
+    index = threadIdx().x + (blockIdx().x - 1) * blockDim().x
+
+    @inbounds if index <= max_idx
+        j, k = divrem(index-1, max_dims_idx)
+        dims_i = CartesianIndices(dims_size)[k+1]
+        dst[index] = src[Tuple(dims_i)..., idx[j+1]...]
```
Comment (CarloLucibello) [inline src/gather.jl:7]:
```suggestion
        dst[index] = src[dims_i, idx[j+1]...]
```
?

File: src/gather.jl (hunk 1 above)
Comment (yuehhua) [reply src/gather.jl:13]: Sure! However, the `gather_check_dims` is generic. It is available for both cpu and gpu. Should I define it in NNlibCUDA? Or it is just fine to put it here?

**Comment by CarloLucibello** [comment]
I see that in [ScatterNNlib](https://github.com/yuehhua/ScatterNNlib.jl/blob/master/src/cuda/gather.jl) the cuda implementation is the same as the cpu one. Wasn't it performing scalar indexin though? Did you compare performance with the implementation here?

**Comment by yuehhua** [comment]
Yeah, they have the same implementation. The key difference here is to put the scalar indexing operator into CUDA kernel. The kernel is crucial to leverage the threads in cuda to accelerate computation. The scalar indexing warning is to index cuarray outside the kernel. I can surely provide some benchmarks between two.

**Comment by yuehhua** [comment]
If I benchmark over cpu version but with cuarray, the scalar operations warning is triggered.

```julia
julia> using NNlib

julia> using BenchmarkTools

julia> using CUDA

julia> T = Float32
Float32

julia> src = cu(T[3, 4, 5, 6, 7]);

julia> index = cu(rand([1,2,3,4,5], 1000, 100));

julia> output = cu(zeros(T, 1000, 100));

julia> NNlib.gather!(output, src, index);
┌ Warning: Performing scalar operations on GPU arrays: This is very slow, consider disallowing these operations with `allowscalar(false)`
└ @ GPUArrays ~/.julia/packages/GPUArrays/4n0iS/src/host/indexing.jl:64

julia> @benchmark NNlib.gather!($output, $src, $index)
BenchmarkTools.Trial: 
  memory estimate:  42.72 MiB
  allocs estimate:  1000000
  --------------
  minimum time:     1.413 s (0.82% GC)
  median time:      1.481 s (1.10% GC)
  mean time:        1.498 s (1.79% GC)
  maximum time:     1.616 s (4.65% GC)
  --------------
  samples:          4
  evals/sample:     1
```

If we have the same setting but use cuda version `gather`, we have better performance.

```julia
julia> using NNlib, NNlibCUDA

julia> using BenchmarkTools

julia> using CUDA

julia> T = Float32
Float32

julia> src = cu(T[3, 4, 5, 6, 7]);

julia> index = cu(rand([1,2,3,4,5], 1000, 100));

julia> output = cu(zeros(T, 1000, 100));

julia> NNlib.gather!(output, src, index);

julia> @benchmark NNlib.gather!($output, $src, $index)
BenchmarkTools.Trial: 
  memory estimate:  272 bytes
  allocs estimate:  10
  --------------
  minimum time:     1.713 μs (0.00% GC)
  median time:      2.633 μs (0.00% GC)
  mean time:        2.600 μs (0.00% GC)
  maximum time:     10.720 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     10
```

The scale difference is 1.5 s and 2.6  μs.

**Comment by CarloLucibello** [comment]
I just don't understand why the gather gpu benchmark in ScatterNNlib shows good performance while the code I see here
https://github.com/yuehhua/ScatterNNlib.jl/blob/master/src/cuda/gather.jl
performs scalar indexing so should be slow

**Comment by CarloLucibello** [comment]
anyways, this pr looks good, merging it

**Comment by CarloLucibello** [comment]
@yuehhua are there any more missing features preventing NNlib's scatter/gather from being used in GeometricFlux?

**Pull request merged**